### DISCUSSION TITLE
Problems forwarding ports

### DISCUSSION THREAD

**Discussion by fugufish** [community/community#24961] [Codespaces]
<p>I have my workspace to forward a particular port for a persistent tcp connection. It works initially, but as soon as I shut down or restart the server the forwarded port disappears from the forward list (despite it being listed in my .devcontainer) and cannot be re-forwarded reliably. I can re-add the port, but it will choose a random port on my local machine to forward to, ex 2323 → 500123 but this only works if the server is not running. If the server is running before I try to forward the port, it will not allow it to happen</p>
<p>pertinent logs:</p>
<p>after initial server shutdown:</p>
<pre><code class="lang-auto">[WorkspaceServices I] onSharingChanged: sourcePort=2323 isTCPServerConnectionEstablished=true hasTLSHandshakePassed=false isPublic=false
[LocalPortForwarder[2323] V] stop: stopping local server port=2323 connections=0
[LocalPortForwarder[2323] V] stop: stopped local server port=2323&gt; object #0: undefined
[WorkspaceServices V] openTunnelForSharedServer: remoteAddress={"host":"localhost","port":2323}
[LiveShareTunnel[2323] I] dispose: Stop sharing port.
[WorkspaceServices V] &lt; serverSharing.stopSharing()
[WorkspaceServices V] &gt; serverSharing.sharingStopped: 
[WorkspaceServices V] &gt; serverSharing.stopSharing() succeeded t:55.0
[ActivityMonitor V] Sending activity heartbeat
[ActivityMonitor V] Sending activity heartbeat
[ActivityMonitor V] Sending activity heartbeat
</code></pre>
<p>after this the shared port is permanently removed.</p>
<p>upon restart:</p>
<pre><code class="lang-auto">[RemoteResolver V] showCandidatePort: port=25967 host=127.0.0.1 detail="/go/bin/dlv\u0000debug\u0000/home/vscode/workspace/fos/main.go\u0000--headless=true\u0000--listen=127.0.0.1:25967\u0000--api-version=2\u0000--wd=.\u0000--\u0000portal\u0000start\u0000" shouldShow=true
[RemoteResolver I] tunnelFactory: port=25967 host=127.0.0.1 public=false
[WorkspaceServices I] forwardRemotePort: forwarding port=25967 host=127.0.0.1 public=false
[WorkspaceServices V] &lt; serverSharing.getSharedServers()
[WorkspaceServices V] &gt; serverSharing.getSharedServers() succeeded t:54.0
[WorkspaceServices V] &lt; serverSharing.startSharing()
[WorkspaceServices V] &gt; serverSharing.sharingStarted: 
[WorkspaceServices I] checkForSharedServers: sourcePort=25967 isTCPServerConnectionEstablished=false hasTLSHandshakePassed=false isPublic=false
[WorkspaceServices V] openTunnelForSharedServer: remoteAddress={"host":"127.0.0.1","port":25967}
[WorkspaceServices V] &gt; serverSharing.browseSharingStarted: 
[WorkspaceServices V] &gt; serverSharing.startSharing() succeeded t:54.0
[WorkspaceServices V] forwardRemotePort: getOrStartLocalPortForwarder host=127.0.0.1 port=25967
[WorkspaceServices V] getOrCreateLocalPortForwarder: port=25967 isTCPServerConnectionEstablished=false hasTLSHandshakePassed=false isPublic=false isEnumerable=true usePFS=true
[LocalPortForwarder[25967] I] start: sourcePort=25967 localPort=25967 host=127.0.0.1
[WorkspaceServices V] getOrCreateLocalPortForwarder: server started host=127.0.0.1 port=25967 localAddress={"port":25967,"host":"127.0.0.1"}
[RemoteResolver V] showCandidatePort: port=2323 host=0:0:0:0:0:0:0:0 detail="/home/vscode/workspace/fos/__debug_bin\u0000portal\u0000start\u0000" shouldShow=true
[LiveShareTunnel[25967] I] dispose: Stop sharing port.
[LocalPortForwarder[25967] V] stop: stopping local server port=25967 connections=0
[LocalPortForwarder[25967] V] stop: stopped local server port=25967&gt; object #0: undefined
[WorkspaceServices V] &lt; serverSharing.stopSharing()
[WorkspaceServices V] &gt; serverSharing.sharingStopped: 
[WorkspaceServices V] &gt; serverSharing.stopSharing() succeeded t:50.0
[ActivityMonitor V] Sending activity heartbeat
</code></pre>
<p>trying to add a port while server is running after server restart</p>
<pre><code class="lang-auto">ActivityMonitor V] Sending activity heartbeat
[RemoteResolver V] showCandidatePort: port=2323 host=0:0:0:0:0:0:0:0 detail="/home/vscode/workspace/fos/__debug_bin\u0000portal\u0000start\u0000" shouldShow=true
[ActivityMonitor V] Sending activity heartbeat
[RemoteResolver I] tunnelFactory: port=2323 host=localhost public=false
[WorkspaceServices I] forwardRemotePort: forwarding port=2323 host=localhost public=false
[WorkspaceServices V] &lt; serverSharing.getSharedServers()
[WorkspaceServices V] &gt; serverSharing.getSharedServers() succeeded t:56.0
[WorkspaceServices V] &lt; serverSharing.startSharing()
[ActivityMonitor V] Sending activity heartbeat
[ActivityMonitor V] Sending activity heartbeat
[ActivityMonitor V] Sending activity heartbeat
</code></pre>
<p>adding a port thereafter without the server running:</p>
<pre><code class="lang-auto">[RemoteResolver I] tunnelFactory: port=2323 host=localhost public=false
[WorkspaceServices I] forwardRemotePort: forwarding port=2323 host=localhost public=false
[WorkspaceServices V] &lt; serverSharing.getSharedServers()
[WorkspaceServices V] &gt; serverSharing.getSharedServers() succeeded t:48.0
[WorkspaceServices V] &lt; serverSharing.startSharing()
[WorkspaceServices V] &gt; serverSharing.sharingStarted: 
[WorkspaceServices I] checkForSharedServers: sourcePort=2323 isTCPServerConnectionEstablished=false hasTLSHandshakePassed=false isPublic=false
[WorkspaceServices V] openTunnelForSharedServer: remoteAddress={"host":"localhost","port":2323}
[WorkspaceServices V] &gt; serverSharing.browseSharingStarted: 
[WorkspaceServices V] &gt; serverSharing.startSharing() succeeded t:58.0
[WorkspaceServices V] forwardRemotePort: getOrStartLocalPortForwarder host=localhost port=2323
[WorkspaceServices V] getOrCreateLocalPortForwarder: port=2323 isTCPServerConnectionEstablished=false hasTLSHandshakePassed=false isPublic=false isEnumerable=true usePFS=true
[LocalPortForwarder[2323] I] start: sourcePort=2323 localPort=50886 host=localhost
[WorkspaceServices V] getOrCreateLocalPortForwarder: server started host=localhost port=2323 localAddress={"port":50886,"host":"127.0.0.1"}
[ActivityMonitor V] Sending activity heartbeat
[ActivityMonitor V] Sending activity heartbeat
</code></pre>
<pre><code class="lang-auto"></code></pre>

**Comment by fugufish** [comment]
<p>I actually was able to manually foward 2323 → 2323 after about 5 or 10 minutes, it makes me think that something internal to the extension is holding onto that port and not closing it, preventing it from being recreated until it times out</p>

**Comment by fugufish** [comment]
<p>any updates on this this is becoming rather bothersome</p>

**Comment by ghost** [comment, marked as answer]
<p>Hi, thanks for your feedback.</p>
<blockquote>
<p>It works initially, but as soon as I shut down or restart the server the forwarded port disappears from the forward list (despite it being listed in my .devcontainer)</p>
</blockquote>
<p>This is a known issue that’s currently being investigated. You can track the progress <a href="https://github.com/microsoft/vscode-remote-release/issues/4550" rel="noopener nofollow ugc">here.</a></p>
<blockquote>
<p>If the server is running before I try to forward the port, it will not allow it to happen</p>
</blockquote>
<p>Also a known issue that sometimes happens, a fixed is planned to rollout in the next codespaces extension release.</p>
